### DISCUSSION TITLE
Access to a protected member

### DISCUSSION THREAD

**Discussion by MikoShoiDev** [pytest-dev/pytest#12046] [General]
**Description:**
Developing frameworks (like pytest) is challenging. There are probably different APIs you'd like to expose to end-user, plugin developers and different API you'd like to use for internal purposes. From my understanding (please, correct me if I'm wrong) pytest uses underscore to different what belongs to pytest-team and what's safe to use by rest of the world (end-users/plugin-developers).

**PEP8** describes a single underscore (among others) as an "internal use" marker. 
Not defining what the internal means exactly (project, class, specific scope). 
![image](https://github.com/pytest-dev/pytest/assets/33802596/4e94d83b-890c-4d6c-b8d4-9ee5574ea940)

**Pylint** definition is probably the most popular one
![image](https://github.com/pytest-dev/pytest/assets/33802596/e0927651-b1c3-4df1-a8b4-f0054eb6b1ae)

**Question:** 
Are the following fragments of code phytonic?

![image](https://github.com/pytest-dev/pytest/assets/33802596/0803cced-9567-40e2-b0e4-bb1453af3287)
![image](https://github.com/pytest-dev/pytest/assets/33802596/6e6b6a78-3d43-4515-b6cf-ec652c92496f)

**Comment by bluetech** [comment]
> From my understanding (please, correct me if I'm wrong) pytest uses underscore to different what belongs to pytest-team and what's safe to use by rest of the world (end-users/plugin-developers).

Pytest code has a long history, and such conventions are not always followed. It's generally true that everything with underscore is private, but the inverse is not true. The official public parts of pytest is what's specified in the API reference: https://docs.pytest.org/en/stable/reference/reference.html

**Comment by MikoShoiDev** [comment]
> > From my understanding (please, correct me if I'm wrong) pytest uses underscore to different what belongs to pytest-team and what's safe to use by rest of the world (end-users/plugin-developers).
> 
> The official public parts of pytest is what's specified in the API reference

In that case I'm wondering why use underscores after all? Why not make it all 'public/without underscores' when it's not code but API reference that tells what's public and private? 

Don't get me wrong I understand there are legacy parts no one wants to change. I'm just trying to understand the logic behind it. I've seen similar code in urllib3, logging or django and at the moment I have no clue if that's state of the art or rather "not expected but accepted". The whole (non)public idea seems to be mixed up. 

Reading PEP8 I feel that the 'internal use' refers not to **where** given element is defined and used but to **who** (framework developer / end-user) defines and uses it. Am I correct?

---------------- 

![image](https://github.com/pytest-dev/pytest/assets/33802596/be5e42c6-e678-4524-97b8-e58b8aa80d57)
![image](https://github.com/pytest-dev/pytest/assets/33802596/8beba718-2a8f-45ff-837c-1b90ec5a08c9)
![image](https://github.com/pytest-dev/pytest/assets/33802596/1d4b60ed-e231-4015-ad27-50bc9c9655eb)

**Reply by RonnyPfannschmidt** [reply]
External users are not to use the underscored objects without a clear understanding that their API isn't subject to stability promise's

**Reply by MikoShoiDev** [reply]
I'm not sure what question your reply answers @RonnyPfannschmidt but thanks anyway 😁. 
(I got that part from the screenshot of PEP I've pasted just above your message) .

**Reply by nicoddemus** [reply]
> In that case I'm wondering why use underscores after all? Why not make it all 'public/without underscores' when it's not code but API reference that tells what's public and private?

We have been using `_` for private members for a number of years now, we just did not go through all the code and change all private methods to `_` yet, mostly because nobody got around to it and also to avoid needless breakage in case somebody is actually using it.
